core/solver/multigrid.cpp

```diff
@@ -159,346 +160,342 @@ void handle_list(
+{
+    return static_cast<bool>(static_cast<int>(a) & static_cast<int>(b));
+}
 
-    template <typename VT>
-    void allocate_memory(int level, multigrid::cycle cycle,
-                         size_type current_nrows, size_type next_nrows)
-    {
-        using vec = matrix::Dense<VT>;
-        using norm_vec = matrix::Dense<remove_complex<VT>>;
-
-        auto exec =
-            as<LinOp>(multigrid->get_mg_level_list().at(level))->get_executor();
-        r_list.emplace_back(vec::create(exec, dim<2>{current_nrows, nrhs}));
-        g_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
-        e_list.emplace_back(vec::create(exec, dim<2>{next_nrows, nrhs}));
-        one_list.emplace_back(initialize<vec>({gko::one<VT>()}, exec));
-        neg_one_list.emplace_back(initialize<vec>({-gko::one<VT>()}, exec));
-        if ((cycle == multigrid::cycle::kfcg ||
-             cycle == multigrid::cycle::kgcr) &&
-            level % multigrid->get_parameters().kcycle_base == 0) {
-            kcycle_state.allocate_memory<VT>(level, cycle, current_nrows,
-                                             next_nrows);
-        }
-    }
 
+namespace detail {
+
+
+/**
+ * MultigridState is used to store the necessary cache and run the operation of
+ * all levels.
+ *
+ * @note it should only be used internally
+ */
+struct MultigridState {
+    MultigridState() : nrhs{0} {}
+
+    /**
+     * Generate the cache for later usage.
+     *
+     * @param system_matrix_in  the system matrix
+     * @param multigrid_in  the multigrid information
+     * @param nrhs_in  the number of right hand side
+     */
+    void generate(const LinOp* system_matrix_in, const Multigrid* multigrid_in,
+                  const size_type nrhs_in);
+
+    /**
+     * allocate_memory is a helper function to allocate the memory of one level
+     *
+     * @tparam ValueType  the value type of memory
+     *
+     * @param level  the current level index
+     * @param cycle  the multigrid cycle
+     * @param current_nrows  the number of rows of current fine matrix
+     * @param next_nrows  the number of rows of next coarse matrix
+     */
+    template <typename ValueType>
+    void allocate_memory(int level, multigrid::cycle cycle,
+                         size_type current_nrows, size_type next_nrows);
+
+    /**
+     * run the cycle of the level
+     *
+     * @param cycle  the multigrid cycle
+     * @param level  the current level index
+     * @param matrix  the system matrix of current level
+     * @param b  the right hand side
+     * @param x  the input vectors
+     * @param info  the info of cycle (See cycle_info)
+     */
+    void run_mg_cycle(multigrid::cycle cycle, size_type level,
+                      const std::shared_ptr<const LinOp>& matrix,
+                      const LinOp* b, LinOp* x,
+                      cycle_info info = cycle_info::first_of_cycle |
+                                        cycle_info::end_of_cycle);
```

Comment:
Since this is an internal API, I think we could also write out this default parameter explicitly, otherwise it may be harder to follow the code (also, this only concerns 2 call sites).

Reply:
Do you mean writing 6 or mode without any default?
I guess the later one. there's no call use the default value so we do not need to provide default here